tests/integration/base.py

```diff
@@ -0,0 +1,299 @@
+                "sam",
+                "init",
+                "--no-input",
+                "--location",
+                template_path,
+                "--name",
+                PROJECT_NAME,
+            ]
+            LOG.info(cmdlist)
+            run_command(cmdlist, self.tempdir.name)
+            self.assertTrue(self.cwd.exists())
+
+    class BuildInvokeBase(IntegBase):
+        function_id_by_event: Optional[Dict[str, str]] = None
+        invoke_output: Dict[str, Any]
+        use_container: bool = True
+
+        def _test_build(self):
+            cmdlist = ["sam", "build", "--skip-pull-image", "--debug"]
+            if self.use_container:
+                cmdlist.append("--use-container")
+            LOG.info(cmdlist)
+            result = run_command(cmdlist, self.cwd)
+            self.assertIn("Build Succeeded", str(result.stdout))
+
+        def _test_local_invoke(self):
+            events_path = Path(self.cwd, "events")
+            if not events_path.exists():
+                LOG.info(f"Skip event testing, {events_path} does not exist")
+                return
+            event_files = os.listdir(events_path)
+            for event_file in event_files:
+                if self.function_id_by_event:
+                    cmdlist = [
+                        "sam",
+                        "local",
+                        "invoke",
+                        self.function_id_by_event[event_file],
+                        "-e",
+                        Path("events", event_file),
+                    ]
+                else:
+                    cmdlist = [
+                        "sam",
+                        "local",
+                        "invoke",
+                        "-e",
+                        Path("events", event_file),
+                    ]
+                LOG.info(cmdlist)
+                result = run_command(cmdlist, self.cwd)
+                try:
+                    self.invoke_output = json.loads(result.stdout.decode())
+                except json.decoder.JSONDecodeError:
+                    self.fail(f"Response is not a valid JSON: {result.stdout.decode()}")
+
+        @pytest.mark.flaky(reruns=3)
+        def test_buld_and_invoke(self):
+            self._test_init_template()
+            self._test_build()
+            self._test_local_invoke()
+
+    class HelloWorldWithLocationBuildInvokeBase(BuildInvokeBase):
+        def _test_local_invoke(self):
+            super()._test_local_invoke()
+            self.assertEqual(self.invoke_output["statusCode"], 200)
+            self.assertEqual(
+                self.invoke_output["headers"],
+                {"X-Custom-Header": "application/json", "Content-Type": "application/json",},
+            )
+            body = json.loads(self.invoke_output["body"])
+            self.assertEqual(body["message"], "hello world")
+            # make sure it is an IP address
+            self.assertRegex(body["location"], r"\d+\.\d+\.\d+\.\d+")
```

Comment:
What about IPv6?

I'd prefer to remove the magic regex and check [if it actually is a valid IP](https://docs.python.org/3/library/ipaddress.html).

Reply:
python has it built in, amazing